### PR TITLE
Add equals, hashCode methods to all Authentication implementation class

### LINKED ISSUE #8953: The Equals of the ClientConfigurationData instance does not reach the expected result
**Is your enhancement request related to a problem? Please describe.**

The authentication in ClientConfigurationData does not implement equals and hashcode, resulting in two instances of the same configuration not being equal.

**Describe the solution you'd like**

Add equals, hashCode methods to all Authentication implementation classes.

**Describe alternatives you've considered**

In ClientConfigurationData in the equals, hashCode method to ignore the field authentication.

**Additional context**

### PR DESCRIPTION
**Pull request by jianyun8023**
Fixes #8953 

### Motivation

Add equals, hashCode methods to all Authentication implementation classes.

### Modifications

*all Authentication implementation class.*

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

This change added tests and can be verified as follows:
  - *Add Authentication implementation class equals test*

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API: (**yes** / no)
  - The schema: (yes / **no** / don't know)
  - The default values of configurations: (yes / **no**)
  - The wire protocol: (yes / **no**)
  - The rest endpoints: (yes / **no**)
  - The admin cli options: (yes / **no**)
  - Anything that affects deployment: (yes / **no** / don't know)

### Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / **not documented**)
  - If a feature is not applicable for documentation, explain why?
  - If a feature is not documented yet in this PR, please create a followup issue for adding the documentation

### CODE REVIEW

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/auth/AuthenticationToken.java [hunk 1]
```diff
@@ -98,4 +100,23 @@ public void start() throws PulsarClientException {
...
+            return true;
+        }
+        if (!(o instanceof AuthenticationToken)) {
+            return false;
+        }
+        AuthenticationToken that = (AuthenticationToken) o;
+        if (tokenSupplier == that.tokenSupplier){
+            return true;
+        }
+        return Objects.equals(tokenSupplier.get(), that.tokenSupplier.get());
```
Comment (merlimat) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/auth/AuthenticationToken.java:115]: This does not look correct. The same supplier is allowed to give each time a different token.

Comment (jianyun8023) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/auth/AuthenticationToken.java:115]: I don't think it's good that this class is designed to make it impossible to compare. Do you have any suggestions?

Comment (merlimat) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/auth/AuthenticationToken.java:115]: I think in most of these authentication classes the equality does not make sense indeed. The classes are meant to be dynamically create and can contain complex state, (eg: opening connections to other systems).

**Comment by codelipenghui** [comment]
@jianyun8023 Could you please take a look at the last comment?

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/auth/AuthenticationToken.java (hunk 1 above)
Comment (jianyun8023) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/auth/AuthenticationToken.java:115]: Your suggestion is correct, I will modify the content here.

**Comment by codelipenghui** [comment]
ping @merlimat Could you please help review this PR again?

**Review by sijie — APPROVED** [review]

**Comment by sijie** [comment]
@jianyun8023 Can you rebase this pull request?

**Review by jiazhai — APPROVED** [review]

**Comment by sijie** [comment]
@jianyun8023 Can you rebase to the latest master?

**Comment by zymap** [comment]
@jianyun8023 Could you please merge the master code? We have some fixes in the master. Thank you.

**Comment by zymap** [comment]
ping @jianyun8023

**Comment by zymap** [comment]
```
Error:  Medium: Inconsistent synchronization of org.apache.pulsar.client.impl.auth.AuthenticationAthenz.privateKey; locked 57% of time [org.apache.pulsar.client.impl.auth.AuthenticationAthenz, org.apache.pulsar.client.impl.auth.AuthenticationAthenz, org.apache.pulsar.client.impl.auth.AuthenticationAthenz, org.apache.pulsar.client.impl.auth.AuthenticationAthenz, org.apache.pulsar.client.impl.auth.AuthenticationAthenz, org.apache.pulsar.client.impl.auth.AuthenticationAthenz, org.apache.pulsar.client.impl.auth.AuthenticationAthenz] Unsynchronized access at AuthenticationAthenz.java:[line 51]Unsynchronized access at AuthenticationAthenz.java:[line 51]Unsynchronized access at AuthenticationAthenz.java:[line 51]Synchronized access at AuthenticationAthenz.java:[line 181]Synchronized access at AuthenticationAthenz.java:[line 143]Synchronized access at AuthenticationAthenz.java:[line 146]Synchronized access at AuthenticationAthenz.java:[line 141] IS2_INCONSISTENT_SYNC
Error:  Medium: Inconsistent synchronization of org.apache.pulsar.client.impl.auth.AuthenticationAthenz.roleHeader; locked 66% of time [org.apache.pulsar.client.impl.auth.AuthenticationAthenz, org.apache.pulsar.client.impl.auth.AuthenticationAthenz, org.apache.pulsar.client.impl.auth.AuthenticationAthenz, org.apache.pulsar.client.impl.auth.AuthenticationAthenz, org.apache.pulsar.client.impl.auth.AuthenticationAthenz, org.apache.pulsar.client.impl.auth.AuthenticationAthenz, org.apache.pulsar.client.impl.auth.AuthenticationAthenz, org.apache.pulsar.client.impl.auth.AuthenticationAthenz, org.apache.pulsar.client.impl.auth.AuthenticationAthenz] Unsynchronized access at AuthenticationAthenz.java:[line 51]Unsynchronized access at AuthenticationAthenz.java:[line 51]Unsynchronized access at AuthenticationAthenz.java:[line 51]Synchronized access at AuthenticationAthenz.java:[line 160]Synchronized access at AuthenticationAthenz.java:[line 161]Synchronized access at AuthenticationAthenz.java:[line 88]Synchronized access at AuthenticationAthenz.java:[line 88]Synchronized access at AuthenticationAthenz.java:[line 100]Synchronized access at AuthenticationAthenz.java:[line 100] IS2_INCONSISTENT_SYNC
Error:  Medium: Inconsistent synchronization of org.apache.pulsar.client.impl.auth.AuthenticationAthenz.roleToken; locked 57% of time [org.apache.pulsar.client.impl.auth.AuthenticationAthenz, org.apache.pulsar.client.impl.auth.AuthenticationAthenz, org.apache.pulsar.client.impl.auth.AuthenticationAthenz, org.apache.pulsar.client.impl.auth.AuthenticationAthenz, org.apache.pulsar.client.impl.auth.AuthenticationAthenz, org.apache.pulsar.client.impl.auth.AuthenticationAthenz, org.apache.pulsar.client.impl.auth.AuthenticationAthenz] Unsynchronized access at AuthenticationAthenz.java:[line 51]Unsynchronized access at AuthenticationAthenz.java:[line 51]Unsynchronized access at AuthenticationAthenz.java:[line 51]Synchronized access at AuthenticationAthenz.java:[line 88]Synchronized access at AuthenticationAthenz.java:[line 98]Synchronized access at AuthenticationAthenz.java:[line 100]Synchronized access at AuthenticationAthenz.java:[line 107] IS2_INCONSISTENT_SYNC
Error:  Medium: Inconsistent synchronization of org.apache.pulsar.client.impl.auth.AuthenticationAthenz.tenantDomain; locked 50% of time [org.apache.pulsar.client.impl.auth.AuthenticationAthenz, org.apache.pulsar.client.impl.auth.AuthenticationAthenz, org.apache.pulsar.client.impl.auth.AuthenticationAthenz, org.apache.pulsar.client.impl.auth.AuthenticationAthenz, org.apache.pulsar.client.impl.auth.AuthenticationAthenz, org.apache.pulsar.client.impl.auth.AuthenticationAthenz] Unsynchronized access at AuthenticationAthenz.java:[line 51]Unsynchronized access at AuthenticationAthenz.java:[line 51]Unsynchronized access at AuthenticationAthenz.java:[line 51]Synchronized access at AuthenticationAthenz.java:[line 181]Synchronized access at AuthenticationAthenz.java:[line 183]Synchronized access at AuthenticationAthenz.java:[line 136] IS2_INCONSISTENT_SYNC
Error:  Medium: Inconsistent synchronization of org.apache.pulsar.client.impl.auth.AuthenticationAthenz.tenantService; locked 50% of time [org.apache.pulsar.client.impl.auth.AuthenticationAthenz, org.apache.pulsar.client.impl.auth.AuthenticationAthenz, org.apache.pulsar.client.impl.auth.AuthenticationAthenz, org.apache.pulsar.client.impl.auth.AuthenticationAthenz, org.apache.pulsar.client.impl.auth.AuthenticationAthenz, org.apache.pulsar.client.impl.auth.AuthenticationAthenz] Unsynchronized access at AuthenticationAthenz.java:[line 51]Unsynchronized access at AuthenticationAthenz.java:[line 51]Unsynchronized access at AuthenticationAthenz.java:[line 51]Synchronized access at AuthenticationAthenz.java:[line 181]Synchronized access at AuthenticationAthenz.java:[line 183]Synchronized access at AuthenticationAthenz.java:[line 137] IS2_INCONSISTENT_SYNC
```

@jianyun8023, please take a look at the CI failure. Looks like introduce by this PR.

**Comment by zymap** [comment]
ping @jianyun8023

**Comment by zymap** [comment]
ping @jianyun8023

**Comment by zymap** [comment]
push to the next release

**Comment by jianyun8023** [comment]
> Error:  Medium: Inconsistent synchronization of org.apache.pulsar.client.impl.auth.AuthenticationAthenz.roleToken; locked 57% of time [org.apache.pulsar.client.impl.auth.AuthenticationAthenz, org.apache.pulsar.client.impl.auth.AuthenticationAthenz, org.apache.pulsar.client.impl.auth.AuthenticationAthenz, org.apache.pulsar.client.impl.auth.AuthenticationAthenz, org.apache.pulsar.client.impl.auth.AuthenticationAthenz, org.apache.pulsar.client.impl.auth.AuthenticationAthenz, org.apache.pulsar.client.impl.auth.AuthenticationAthenz] Unsynchronized access at AuthenticationAthenz.java:[line 51]Unsynchronized access at AuthenticationAthenz.java:[line 51]Unsynchronized access at AuthenticationAthenz.java:[line 51]Synchronized access at AuthenticationAthenz.java:[line 88]Synchronized access at AuthenticationAthenz.java:[line 98]Synchronized access at AuthenticationAthenz.java:[line 100]Synchronized access at AuthenticationAthenz.java:[line 107] IS2_INCONSISTENT_SYNC

My changes didn't affect here, suggest a new issue to deal with it.

**Comment by eolivelli** [comment]
this patch is too dangerous for a port to 2.7.2.
let's do it in master branch and give time to settle.

**Pull request closed without merging**